src/test/lib/logical_query_plan/join_node_test.cpp

```diff
@@ -208,69 +208,157 @@ TEST_F(JoinNodeTest, IsColumnNullableWithOuterJoin) {
+  EXPECT_EQ(_mock_node_a->non_trivial_functional_dependencies().size(), 1);
+  EXPECT_EQ(_mock_node_a->non_trivial_functional_dependencies().at(0), fd_a);
+  _mock_node_b->set_non_trivial_functional_dependencies({fd_x});
+  EXPECT_EQ(_mock_node_b->non_trivial_functional_dependencies().size(), 1);
+  EXPECT_EQ(_mock_node_b->non_trivial_functional_dependencies().at(0), fd_x);
+
+  // Tests
+  for (const auto join_mode : {JoinMode::Semi, JoinMode::AntiNullAsTrue, JoinMode::AntiNullAsFalse}) {
+    // clang-format off
+    const auto join_node =
+        JoinNode::make(join_mode, equals_(_t_a_a, _t_b_y),
+                       _mock_node_a,
+                       _mock_node_b);
+    // clang-format on
 
-  // clang-format off
-  const auto inner_join_node =
-  JoinNode::make(JoinMode::Inner, equals_(join_column_a, join_column_x),
-    mock_node_a,
-    mock_node_b);
-
-  const auto left_join_node =
-  JoinNode::make(JoinMode::Left, equals_(join_column_a, join_column_x),
-    mock_node_a,
-    mock_node_b);
-
-  const auto right_join_node =
-  JoinNode::make(JoinMode::Right, equals_(join_column_a, join_column_x),
-    mock_node_a,
-    mock_node_b);
-
-  const auto full_outer_join_node =
-  JoinNode::make(JoinMode::FullOuter, equals_(join_column_a, join_column_x),
-    mock_node_a,
-    mock_node_b);
-  // clang-format on
+    // We do not want JoinNode to return the FDs of the right input node
+    const auto& non_trivial_fds = join_node->non_trivial_functional_dependencies();
+    EXPECT_EQ(non_trivial_fds.size(), 1);
+    EXPECT_EQ(non_trivial_fds.at(0), fd_a);
+  }
+}
 
-  // Prerequisite
-  EXPECT_EQ(mock_node_a->functional_dependencies().size(), 1);
-  EXPECT_EQ(mock_node_a->functional_dependencies().at(0), fd_ab);
-  EXPECT_EQ(mock_node_b->functional_dependencies().size(), 1);
-  EXPECT_EQ(mock_node_b->functional_dependencies().at(0), fd_xy);
+TEST_F(JoinNodeTest, FunctionalDependenciesForwarding) {
+  const auto fd_a = FunctionalDependency{{_t_a_a}, {_t_a_b}};
+  const auto fd_x = FunctionalDependency{{_t_b_x}, {_t_b_y}};
+  const auto generated_fd_b_c = FunctionalDependency{{_t_a_b, _t_a_c}, {_t_a_a}};
+  const auto generated_fd_y = FunctionalDependency{{_t_b_y}, {_t_b_x}};
+  const auto join_column_a = _t_a_a;
+  const auto join_column_x = _t_b_x;
+
+  for (const auto join_mode :
+       {JoinMode::Inner, JoinMode::Left, JoinMode::Right, JoinMode::FullOuter, JoinMode::Cross}) {
+    auto join_node = std::shared_ptr<JoinNode>();
+    if (join_mode == JoinMode::Cross) {
+      join_node = JoinNode::make(JoinMode::Cross, _mock_node_a, _mock_node_b);
+    } else {
+      join_node = JoinNode::make(join_mode, equals_(join_column_a, join_column_x), _mock_node_a, _mock_node_b);
+    }
+
+    // Left input Node has non-trivial FDs
+    _mock_node_a->set_key_constraints({});
+    _mock_node_b->set_key_constraints({});
+    _mock_node_a->set_non_trivial_functional_dependencies({fd_a});
+    _mock_node_b->set_non_trivial_functional_dependencies({});
```

Comment:
This looks like you are resetting the test a couple of times. Split this into several test cases to improve isolation?

Reply:
Legit point. I split it into four tests.